Releases 1.1.0+ cause Terraform to crash

After upgrading our provider from 1.0.0 to 1.3.0 we encountered the error below. Testing using releases 1.1.0 and 1.2.0 and found the same result.

## Resources

```
resource "sumologic_collector" "cluster" {
  name        = "${var.cluster}"
  description = "Collector for the ${var.cluster} cluster"
}

resource "sumologic_http_source" "kubernetes" {
  name         = "kubernetes"
  collector_id = "${sumologic_collector.cluster.id}"
}
```

## TF output
```
> terraform plan -target module.sumologic

Refreshing Terraform state in-memory prior to plan...
The refreshed state will be used to calculate this plan, but will not be
persisted to local or remote state storage.

module.sumologic.sumologic_collector.cluster: Refreshing state... [id=135201152]

Error: rpc error: code = Unavailable desc = transport is closing


panic: runtime error: invalid memory address or nil pointer dereference
2019-11-03T02:08:47.996Z [DEBUG] plugin.terraform-provider-sumologic_v1.3.0: [signal SIGSEGV: segmentation violation code=0x1 addr=0x48 pc=0xd549db]
2019-11-03T02:08:47.996Z [DEBUG] plugin.terraform-provider-sumologic_v1.3.0:
2019-11-03T02:08:47.996Z [DEBUG] plugin.terraform-provider-sumologic_v1.3.0: goroutine 23 [running]:
2019-11-03T02:08:47.996Z [DEBUG] plugin.terraform-provider-sumologic_v1.3.0: github.com/SumoLogic/sumologic-terraform-provider/sumologic.(*Client).Get(0xc0000cf1d0, 0xc0000ea2c0, 0x17, 0x1, 0x1, 0xc0000ea2c0, 0x17, 0x18, 0x1a468a0, 0x0)
2019-11-03T02:08:47.996Z [DEBUG] plugin.terraform-provider-sumologic_v1.3.0: 	/Users/rmiller/Development/terraform/sumologic-terraform-provider/sumologic/sumologic_client.go:186 +0x1cb
2019-11-03T02:08:47.996Z [DEBUG] plugin.terraform-provider-sumologic_v1.3.0: github.com/SumoLogic/sumologic-terraform-provider/sumologic.(*Client).GetCollector(0xc0000cf1d0, 0x80f0180, 0x80f0180, 0x0, 0x0)
2019-11-03T02:08:47.996Z [DEBUG] plugin.terraform-provider-sumologic_v1.3.0: 	/Users/rmiller/Development/terraform/sumologic-terraform-provider/sumologic/sumologic_collectors.go:9 +0xb8
2019-11-03T02:08:47.996Z [DEBUG] plugin.terraform-provider-sumologic_v1.3.0: github.com/SumoLogic/sumologic-terraform-provider/sumologic.resourceSumologicCollectorExists(0xc00026c770, 0x10063c0, 0xc0000cf1d0, 0xc00026c770, 0x0, 0x0)
2019-11-03T02:08:47.996Z [DEBUG] plugin.terraform-provider-sumologic_v1.3.0: 	/Users/rmiller/Development/terraform/sumologic-terraform-provider/sumologic/resource_sumologic_collector.go:162 +0x1b2
2019-11-03T02:08:47.996Z [DEBUG] plugin.terraform-provider-sumologic_v1.3.0: github.com/hashicorp/terraform/helper/schema.(*Resource).RefreshWithoutUpgrade(0xc0003e6080, 0xc0006201e0, 0x10063c0, 0xc0000cf1d0, 0xc000358360, 0xc0006201e0, 0x0)
2019-11-03T02:08:47.996Z [DEBUG] plugin.terraform-provider-sumologic_v1.3.0: 	/Users/rmiller/.go/pkg/mod/github.com/hashicorp/terraform@v0.12.0/helper/schema/resource.go:432 +0x24d
2019-11-03T02:08:47.996Z [DEBUG] plugin.terraform-provider-sumologic_v1.3.0: github.com/hashicorp/terraform/helper/plugin.(*GRPCProviderServer).ReadResource(0xc0003d0040, 0x1257520, 0xc00009a930, 0xc0001d2400, 0xc0003d0040, 0xc00009a930, 0xc000320bd0)
2019-11-03T02:08:47.996Z [DEBUG] plugin.terraform-provider-sumologic_v1.3.0: 	/Users/rmiller/.go/pkg/mod/github.com/hashicorp/terraform@v0.12.0/helper/plugin/grpc_provider.go:496 +0x334
2019-11-03T02:08:47.996Z [DEBUG] plugin.terraform-provider-sumologic_v1.3.0: github.com/hashicorp/terraform/internal/tfplugin5._Provider_ReadResource_Handler(0xfd6980, 0xc0003d0040, 0x1257520, 0xc00009a930, 0xc0006200a0, 0x0, 0x1257520, 0xc00009a930, 0xc000322000, 0xbf)
2019-11-03T02:08:47.996Z [DEBUG] plugin.terraform-provider-sumologic_v1.3.0: 	/Users/rmiller/.go/pkg/mod/github.com/hashicorp/terraform@v0.12.0/internal/tfplugin5/tfplugin5.pb.go:2983 +0x23e
2019-11-03T02:08:47.997Z [DEBUG] plugin.terraform-provider-sumologic_v1.3.0: google.golang.org/grpc.(*Server).processUnaryRPC(0xc0004e0480, 0x1262f20, 0xc000001500, 0xc000268000, 0xc00025e1b0, 0x1a1b5b0, 0x0, 0x0, 0x0)
2019-11-03T02:08:47.997Z [DEBUG] plugin.terraform-provider-sumologic_v1.3.0: 	/Users/rmiller/.go/pkg/mod/google.golang.org/grpc@v1.18.0/server.go:966 +0x470
2019-11-03T02:08:47.997Z [DEBUG] plugin.terraform-provider-sumologic_v1.3.0: google.golang.org/grpc.(*Server).handleStream(0xc0004e0480, 0x1262f20, 0xc000001500, 0xc000268000, 0x0)
2019-11-03T02:08:47.997Z [DEBUG] plugin.terraform-provider-sumologic_v1.3.0: 	/Users/rmiller/.go/pkg/mod/google.golang.org/grpc@v1.18.0/server.go:1245 +0xd25
2019-11-03T02:08:47.997Z [DEBUG] plugin.terraform-provider-sumologic_v1.3.0: google.golang.org/grpc.(*Server).serveStreams.func1.1(0xc0003e4190, 0xc0004e0480, 0x1262f20, 0xc000001500, 0xc000268000)
2019-11-03T02:08:47.997Z [DEBUG] plugin.terraform-provider-sumologic_v1.3.0: 	/Users/rmiller/.go/pkg/mod/google.golang.org/grpc@v1.18.0/server.go:685 +0x9f
2019-11-03T02:08:47.997Z [DEBUG] plugin.terraform-provider-sumologic_v1.3.0: created by google.golang.org/grpc.(*Server).serveStreams.func1
2019-11-03T02:08:47.997Z [DEBUG] plugin.terraform-provider-sumologic_v1.3.0: 	/Users/rmiller/.go/pkg/mod/google.golang.org/grpc@v1.18.0/server.go:683 +0xa1
2019-11-03T02:08:47.998Z [DEBUG] plugin: plugin process exited: path=/root/.terraform.d/plugins/terraform-provider-sumologic_v1.3.0 pid=5918 error="exit status 2"
2019/11/03 02:08:47 [ERROR] module.sumologic: eval: *terraform.EvalRefresh, err: rpc error: code = Unavailable desc = transport is closing
2019/11/03 02:08:47 [ERROR] module.sumologic: eval: *terraform.EvalSequence, err: rpc error: code = Unavailable desc = transport is closing
2019/11/03 02:08:47 [TRACE] [walkRefresh] Exiting eval tree: module.sumologic.sumologic_collector.cluster
2019/11/03 02:08:47 [TRACE] vertex "module.sumologic.sumologic_collector.cluster": visit complete
2019/11/03 02:08:47 [TRACE] vertex "module.sumologic.sumologic_collector.cluster": dynamic subgraph encountered errors
2019/11/03 02:08:47 [TRACE] vertex "module.sumologic.sumologic_collector.cluster": visit complete
2019/11/03 02:08:47 [TRACE] dag/walk: upstream of "module.sumologic.sumologic_http_source.kubernetes" errored, so skipping
2019/11/03 02:08:47 [TRACE] dag/walk: upstream of "module.sumologic.output.kubernetes_endpoint" errored, so skipping
2019/11/03 02:08:47 [TRACE] dag/walk: upstream of "module.sumologic.local_file.endpoints" errored, so skipping
2019/11/03 02:08:47 [TRACE] dag/walk: upstream of "provider.local (close)" errored, so skipping
2019/11/03 02:08:47 [TRACE] dag/walk: upstream of "provider.sumologic (close)" errored, so skipping
2019/11/03 02:08:48 [TRACE] dag/walk: upstream of "root" errored, so skipping
2019-11-03T02:08:48.000Z [DEBUG] plugin: plugin exited
2019-11-03T02:08:48.003Z [DEBUG] plugin: plugin process exited: path=/usr/local/bin/terraform pid=5883
2019-11-03T02:08:48.003Z [DEBUG] plugin: plugin exited
2019-11-03T02:08:48.006Z [DEBUG] plugin: plugin process exited: path=/conf/backing-services/.terraform/plugins/linux_amd64/terraform-provider-local_v1.4.0_x4 pid=5928
2019-11-03T02:08:48.006Z [DEBUG] plugin: plugin exited



!!!!!!!!!!!!!!!!!!!!!!!!!!! TERRAFORM CRASH !!!!!!!!!!!!!!!!!!!!!!!!!!!!

```

--- a/sumologic/provider.go
+++ b/sumologic/provider.go
@@ -69,13 +69,16 @@ func providerConfigure(d *schema.ResourceData) (interface{}, error) {
 	if accessId == "" {
 		msg = "sumologic provider: access_id should be set;"
 	}
+
 	if accessKey == "" {
 		msg = fmt.Sprintf("%s access_key should be set; ", msg)
 	}
+
+	if environment == "" && baseUrl == "" {
+		msg = fmt.Sprintf("%s make sure one of environment or base_url is set", msg)
+	}
+
 	if msg != "" {
-		if environment == "" {
-			msg = fmt.Sprintf("%s make sure environment is set", msg)
-		}
 		return nil, errors.New(msg)
 	}
 

--- a/sumologic/sumologic_client.go
+++ b/sumologic/sumologic_client.go
@@ -60,6 +60,7 @@ func (s *Client) PostWithCookies(urlPath string, payload interface{}) ([]byte, [
 	if err != nil {
 		return nil, nil, err
 	}
+	defer resp.Body.Close()
 
 	respCookie := resp.Cookies()
 
@@ -100,6 +101,7 @@ func (s *Client) GetWithCookies(urlPath string, cookies []*http.Cookie) ([]byte,
 	if err != nil {
 		return nil, "", err
 	}
+	defer resp.Body.Close()
 
 	d, err := ioutil.ReadAll(resp.Body)
 	if err != nil {
@@ -126,12 +128,15 @@ func (s *Client) Post(urlPath string, payload interface{}) ([]byte, error) {
 
 	<-rateLimiter
 	resp, err := s.httpClient.Do(req)
-
 	if err != nil {
 		return nil, err
 	}
+	defer resp.Body.Close()
 
-	d, _ := ioutil.ReadAll(resp.Body)
+	d, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return nil, err
+	}
 
 	if resp.StatusCode >= 400 {
 		return nil, errors.New(string(d))
@@ -158,12 +163,15 @@ func (s *Client) Put(urlPath string, payload interface{}) ([]byte, error) {
 
 	<-rateLimiter
 	resp, err := s.httpClient.Do(req)
-
 	if err != nil {
 		return nil, err
 	}
+	defer resp.Body.Close()
 
-	d, _ := ioutil.ReadAll(resp.Body)
+	d, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return nil, err
+	}
 
 	if resp.StatusCode >= 400 {
 		return nil, errors.New(string(d))
@@ -181,9 +189,16 @@ func (s *Client) Get(urlPath string) ([]byte, string, error) {
 	req.SetBasicAuth(s.AccessID, s.AccessKey)
 
 	<-rateLimiter
-	resp, _ := s.httpClient.Do(req)
+	resp, err := s.httpClient.Do(req)
+	if err != nil {
+		return nil, "", err
+	}
+	defer resp.Body.Close()
 
-	d, _ := ioutil.ReadAll(resp.Body)
+	d, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return nil, "", err
+	}
 
 	if resp.StatusCode == 404 {
 		return nil, "", nil
@@ -203,9 +218,16 @@ func (s *Client) Delete(urlPath string) ([]byte, error) {
 	req.SetBasicAuth(s.AccessID, s.AccessKey)
 
 	<-rateLimiter
-	resp, _ := s.httpClient.Do(req)
+	resp, err := s.httpClient.Do(req)
+	if err != nil {
+		return nil, err
+	}
+	defer resp.Body.Close()
 
-	d, _ := ioutil.ReadAll(resp.Body)
+	d, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return nil, err
+	}
 
 	if resp.StatusCode >= 400 {
 		return nil, errors.New(string(d))

--- a/sumologic/sumologic_client.go
+++ b/sumologic/sumologic_client.go
@@ -4,6 +4,7 @@ import (
 	"bytes"
 	"encoding/json"
 	"errors"
+	"fmt"
 	"io/ioutil"
 	"net/http"
 	"net/url"
@@ -243,10 +244,19 @@ func NewClient(accessID, accessKey, environment, base_url string) (*Client, erro
 		httpClient:  http.DefaultClient,
 		Environment: environment,
 	}
+
 	if base_url == "" {
-		base_url = endpoints[client.Environment]
+		var found bool
+		base_url, found = endpoints[client.Environment]
+		if !found {
+			return nil, fmt.Errorf("could not find endpoint for environment %s", environment)
+		}
 	}
-	client.BaseURL, _ = url.Parse(base_url)
+	parsed, err := url.Parse(base_url)
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse base_url %s", base_url)
+	}
+	client.BaseURL = parsed
 
 	return &client, nil
 }

